codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -84,14 +87,11 @@
 
     private final ChannelHandler inboundStreamHandler;
     private final ChannelHandler upgradeStreamHandler;
+    private final Queue<AbstractHttp2StreamChannel> readPendingQueue = new ArrayDeque<AbstractHttp2StreamChannel>(8);
```

Comment:
Since removal is no longer needed and there is no longer a check to see if it is already added, the original list could have become singly-linked, which would be much simpler as well.

Using an ArrayDeque does have a "high water mark" problem, where it never shrinks in size. If there is a large batch of concurrent RPCs and then a much smaller amount thereafter (let's say the connection becomes idle), we'll continue having a "large" array here. This may be "okay" today, as we have the same problem elsewhere (like DefaultHttp2Connection's streamMap). It may be possible to claim we intend MAX_CONCURRENT_STREAMS to guard against that.

Reply:
Yeah I was thinking of this as well and then decided that we already have bound here. But I can also revert this part if you think we should better not do it

Reply:
@ejona86 also we could also limit the max queue size and just call "fireChannelReadComplete" for everything contained so far and empty it. After this we just fill it again.

Reply:
Ooh. Interesting idea. SGTM. Even with a max value of 100 it'd be highly unlikely we'd receive a frame for each of 100 streams. Go for it! (I'd be okay with a value as low as 20... maybe 32 so it is a power-of-two. But we may just choose 100 to match the common value for MAX_CONCURRENT_STREAMS)